KSES: Add support for rgb(a) color.

t-hamano opened the pull request:
Trac ticket: https://core.trac.wordpress.org/ticket/56391

Related gutenberg issue: https://github.com/WordPress/gutenberg/issues/39402
Related gutenberg PR: https://github.com/WordPress/gutenberg/pull/39488

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

t-hamano committed:
KSES: Add support for rgba background-color.

t-hamano committed:
Fix lint error

peterwilsoncc commented on src/wp-includes/kses.php:
```diff
@@ -2395,6 +2395,14 @@ function safecss_filter_attr( $css, $deprecated = '' ) {
+	 */
+	$css_rgba_color_data_types = array(
+		'background-color',
+	);
```
```suggestion
	$css_color_data_types = array(
		'color',
		'border-color',
		/* ETC ETC for all of the items that support color values in the safe CSS list */
		'background-color',
	);
```

A generic name to allow for other color types to be added later: there are currently specs been written for new color types so it will become a thing sooner rather than later.

peterwilsoncc commented on src/wp-includes/kses.php:
```diff
@@ -2418,6 +2426,7 @@ function safecss_filter_attr( $css, $deprecated = '' ) {
 		$css_test_string = $css_item;
 		$found           = false;
 		$url_attr        = false;
+		$rgba_attr       = false;
```
```suggestion
		$color_attr      = false;
```

And where the variable is used throughout.

peterwilsoncc commented on src/wp-includes/kses.php:
```diff
@@ -2466,6 +2476,13 @@ function safecss_filter_attr( $css, $deprecated = '' ) {
 
+		if ( $found && $rgba_attr ) {
+			$css_value = trim( $parts[1] );
+			if ( preg_match( '/^rgba\((\d{1,3}%?),\s*(\d{1,3}%?),\s*(\d{1,3}%?),\s*(\d*(?:\.\d+)?)\)$/', $css_value ) ) {
```
Is it possible to accept both `rgb` and `rgba`? It will be super confusing to people as to why one is accepted but not the other. `rgb` now accepts a forth argument for transparency so they're essentially the same function. See the [MDN color docs](http://developer.mozilla.org/en-US/docs/Web/CSS/color_value#rgb()).

This will require some changes to the `test_safecss_filter_attr_filtered` data provider to promote rgb to be always allowed.

peterwilsoncc commented on tests/phpunit/tests/kses.php:
```diff
@@ -1110,6 +1110,31 @@ public function data_test_safecss_filter_attr() {
+			array(
+				'css'      => 'background-color: rgba(0, 0, 0, .1)',
+				'expected' => 'background-color: rgba(0, 0, 0, .1)',
+			),
```
There's an test below under the comment "RGBA color values are not allowed." that will now fail.

It would be good to test for some malformed values too.

t-hamano replied:
I am thinking about regular expression and test cases, am I correct in this perception?

✅ `color: rgba(0, 0, 0 / 0.1)`
❌ `color: rgba(0, 0, 0/ 0.1)` (don't allow no space before slash)
❌ `color: rgba(0, 0, 0 /0.1)` (don't allow no space after slash)

✅ `color: rgba(0, 0, 0, 0)`
✅ `color: rgba(0,0,0,0)` (allow no space after comma)
❌ `color: rgba(0 , 0, 0, 0)` (don't allow space before comma)
❌ `color: rgba(0,   0, 0, 0)` (don't allow multiple spaces after comma *it would appear as single space on the comments)

✅ `color: rgba(100%, 0, 0, 0)`
❌ `color: rgba(100px,  0, 0, 0)` (don't allow units other than %)
❌ `color: rgba(red,  0, 0, 0)` (don't allow string value)

According to MDN, I believe the same rules can be applied to both [rgb()](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value/rgb) and [rgba()](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value/rgba).

peterwilsoncc replied:
> I am thinking about regular expression and test cases, am I correct in this perception?

Yes, you are. These are the sort of tests I was suggesting. 

You may also wish to include something containing either a `"` or a `'` as an attempt to close the style tag. The expectation is that kses would return an empty string. 

Thanks so much.

peterwilsoncc reviewed (changes requested):
I've added a few notes inline.

t-hamano committed:
support rgb color

t-hamano commented:
I have added support for RGB and allowed some properties, and updated regular expressions.
And I have tried to cover as many variations of the test as possible. Is this enough?

peterwilsoncc commented on src/wp-includes/kses.php:
```diff
@@ -2395,6 +2395,22 @@ function safecss_filter_attr( $css, $deprecated = '' ) {
+		'border-left-color',
+		'border-top-color',
+
+		'background-color',
```
Also don't forget the shorthand
* `background`
* `border`
* `border-(left|right|top|bottom)`

That's assuming support :)

peterwilsoncc commented on tests/phpunit/tests/kses.php:
```diff
@@ -1110,14 +1110,86 @@ public function data_test_safecss_filter_attr() {
+				'expected' => '',
+			),
+			array(
+				'css'      => 'border-right-color: rgba(0,  0, 0, 0)',
```
I'd expect this to pass and be retained. The format is a bit ugly but it's valid https://jsbin.com/viripuz/1/edit?css,output

peterwilsoncc commented on src/wp-includes/kses.php:
```diff
@@ -2466,6 +2484,13 @@ function safecss_filter_attr( $css, $deprecated = '' ) {
 
+		if ( $found && $color_attr ) {
+			$css_value = trim( $parts[1] );
+			if ( preg_match( '/^rgb[a]?\((\d{1,3}%?),\s?(\d{1,3}%?),\s?(\d{1,3}%?)(,\s?|\s\/\s)(\d*(?:\.\d+)?)\)$/', $css_value ) ) {
```
It might be possible to simplify the regex a lot here.

The goal of kses is to make the HTML (and CSS it contains) safe rather than valid. It's nice if it's valid but that's a problem for other parts of the code base.

You might be able to use `/^rgba?\([\s\d%\/.,]+\)$/` to limit the user to the allowed characters

peterwilsoncc reviewed (commented):
I've added a few more notes inline.

There's a few cases in tests in which the value is valid but poorly formatted that are getting dropped. These ought to be allowed.

I've made a suggestion for much simpler regex too.

t-hamano committed:
Add missing supports

t-hamano committed:
Relaxed rules for regular expression

t-hamano committed:
Update tests

t-hamano commented:
I have relaxed the regex rules and updated the existing tests.
With the relaxed rules, should we reduce the number of test cases a little more?

t-hamano commented:
The problem this PR is trying to solve appears to have been fixed by [Changeset 54117](https://core.trac.wordpress.org/changeset/54117).
Therefore, I would like to close this PR.

t-hamano commented:
> The problem this PR is trying to solve appears to have been fixed by [Changeset 54117](https://core.trac.wordpress.org/changeset/54117).

I realized this wasn't correct. This changeset only allowed CSS variables, so RGB(A) colors are still not allowed. We will reopen this PR and resume work on it.

t-hamano committed:
Merge branch 'trunk' into fix/kses-allow-background-color

t-hamano committed:
Refactoring regular expressions and tests

t-hamano commented:
I started by using the equivalent regular expression from the [colored_parse_rgba_string() function](https://github.com/t-hamano/wordpress-develop/blob/97fe09e734d9900bd2b92eff31df5b20ddfef45b/src/wp-includes/class-wp-duotone.php#L302-L317). This logic should work well for both comma and space value separators. I also overhauled the tests to test as many patterns as possible without overdoing it.

t-hamano commented:
Oh, I re-forked the `wordpress-develop` repo and it seems all my submitted open PRs were closed 😂 I submitted a new PR that is equivalent to this one. https://github.com/WordPress/wordpress-develop/pull/10923
